variable/logical.cpp

```diff
@@ -14,28 +14,40 @@ using namespace scipp::core;
 
 namespace scipp::variable {
 
-static constexpr auto or_op_ = overloaded{
-    [](const auto &var_, const auto &other_) -> bool { return var_ | other_; },
-    dimensionless_unit_check_return};
+static constexpr auto or_op_ =
+    overloaded{[](const auto &var_, const auto &other_) -> bool {
+                 return static_cast<decltype(other_)>(var_) | other_;
+               },
+               dimensionless_unit_check_return};
```

Comment:
Why are these changes required? This should only ever be called with `bool`?

Reply:
MSVC warns about bool and int comparisons, so I'll add a static assert that the LHS/RHS are both bool types and see what falls over

Reply:
Hmm, is it because we are using bit-wise operators, and it casts the bools to ints? Otherwise I would not see why, both inputs should be bool?!

Reply:
https://stackoverflow.com/questions/44565126/how-to-enable-c-warnings-for-bitwise-operators-with-boolean-arguments?

Reply:
This is similar to the problem in rebin, the compiler warning shows a generated lambda with bool / int for example
I suspect fixing one will fix multiple places where MSVC is warning about this

Reply:
and https://stackoverflow.com/questions/24542/using-bitwise-operators-for-booleans-in-c

Reply:
So it's probably worth changing these bitwise operators to logical operators too, since in SCIPP were only concerned about True / False instead of bit manipulation

Reply:
Interestingly:
```python
import numpy as np

a = np.array([True, False])
~a # array([False,  True])
```
That is, numpy behavior does not match Python behavior, so the operators we provide appear to be consistent with numpy (we may just need to implement them differently)? As mentioned on Slack, I suggest to move this to a different issue.